File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/AI/Prompts/Sections/TextSection.cs
﻿using Microsoft.Teams.AI.AI.Models;
using Microsoft.Bot.Builder;
using Microsoft.Teams.AI.AI.Tokenizers;
using Microsoft.Teams.AI.State;

namespace Microsoft.Teams.AI.AI.Prompts.Sections
{
    /// <summary>
    /// A section of text that will be rendered as a message.
    /// </summary>
    public class TextSection : PromptSection
    {
        /// <summary>
        /// Chat Message Content
        /// </summary>
        public readonly string text;

        /// <summary>
        /// Chat Message Role
        /// </summary>
        public readonly ChatRole role;

        private int _length = -1;

        /// <summary>
        /// Creates instance of `TextSection`
        /// </summary>
        /// <param name="text">message text</param>
        /// <param name="role">message role</param>
        /// <param name="tokens">tokens</param>
        /// <param name="required">required</param>
        /// <param name="separator">separator</param>
        /// <param name="prefix">prefix</param>
        public TextSection(string text, ChatRole role, int tokens = -1, bool required = false, string separator = "\n", string prefix = "") : base(tokens, required, separator, prefix)
        {
            this.text = text;
            this.role = role;
        }

        /// <inheritdoc />
        public override async Task<RenderedPromptSection<List<ChatMessage>>> RenderAsMessagesAsync(ITurnContext context, IMemory memory, IPromptFunctions<List<string>> functions, ITokenizer tokenizer, int maxTokens, CancellationToken cancellationToken = default)
        {
            // calculate and cache length
            if (this._length < 0)
            {
                this._length = tokenizer.Encode(this.text).Count;
            }

            List<ChatMessage> messages = new();

            if (this._length > 0)
            {
                messages.Add(new(this.role) { Content = this.text });
            }

            return await Task.FromResult(this.TruncateMessages(messages, tokenizer, maxTokens));
        }
    }
}

File: python/packages/ai/teams/ai/promptsv2/text_section.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from typing import List

from botbuilder.core import TurnContext

from ...state import Memory
from ..tokenizers import Tokenizer
from .message import Message
from .prompt_functions import PromptFunctions
from .prompt_section_base import PromptSectionBase
from .rendered_prompt_section import RenderedPromptSection


class TextSection(PromptSectionBase):
    _text: str
    _role: str
    _length: int

    # pylint: disable=too-many-arguments # No argument can be removed based on the design
    def __init__(
        self,
        text: str,
        role: str,
        tokens: float = -1,
        required: bool = True,
        separator: str = "\n",
        text_prefix: str = "",
    ):
        super().__init__(tokens, required, separator, text_prefix)
        self._text = text
        self._role = role
        self._length = -1

    # pylint: enable=too-many-arguments

    @property
    def text(self):
        return self._text

    @property
    def role(self):
        return self._role

    # pylint: disable=too-many-arguments # No argument can be removed based on the design
    async def render_as_messages(
        self,
        context: TurnContext,
        memory: Memory,
        functions: PromptFunctions,
        tokenizer: Tokenizer,
        max_tokens: int,
    ) -> RenderedPromptSection[List[Message]]:
        # Calculate and cache length
        if self._length < 0:
            self._length = len(tokenizer.encode(self.text))

        # Return output
        messages: List[Message] = [Message(self.role, self.text)] if self._length > 0 else []
        return self._return_messages(messages, self._length, tokenizer, max_tokens)

    # pylint: enable=too-many-arguments
